numpy/core/src/multiarray/compiled_base.c

```diff
@@ -302,193 +302,152 @@ arr_digitize(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
+    values = NULL;
+    array = (PyArrayObject *)array0;
 
-    sameshape = 1;
-    if (PyArray_NDIM(amask) == PyArray_NDIM(ainput)) {
-        for (k = 0; k < PyArray_NDIM(amask); k++) {
-            if (PyArray_DIMS(amask)[k] != PyArray_DIMS(ainput)[k]) {
-                sameshape = 0;
-            }
-        }
+    if (!PyArray_Check(array)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "place: first argument must "
+                        "be an array");
+        return NULL;
     }
-    else {
-        /* Test to see if amask is 1d */
-        if (PyArray_NDIM(amask) != 1) {
-            sameshape = 0;
-        }
-        else if ((PyArray_SIZE(ainput)) != PyArray_SIZE(amask)) {
-            sameshape = 0;
+
+    if (!PyArray_ISCONTIGUOUS(array)) {
+        PyArrayObject *obj;
+
+        dtype = PyArray_DESCR(array);
+        Py_INCREF(dtype);
+        obj = (PyArrayObject *)PyArray_FromArray(array, dtype,
+                                NPY_ARRAY_CARRAY | NPY_ARRAY_UPDATEIFCOPY);
+        if (obj != array) {
+            copied = 1;
         }
+        array = obj;
     }
-    if (!sameshape) {
-        PyErr_SetString(PyExc_TypeError,
-                        "mask array must be 1-d or same shape as input array");
+
+    max_item = PyArray_SIZE(array);
+    dest = PyArray_DATA(array);
+    chunk = PyArray_DESCR(array)->elsize;
+    mask = (PyArrayObject *)PyArray_FROM_OTF(mask0, NPY_BOOL,
+                                NPY_ARRAY_CARRAY | NPY_ARRAY_FORCECAST);
+    if (mask == NULL) {
         goto fail;
     }
 
-    avals = (PyArrayObject *)PyArray_FromObject(vals,
-                                        PyArray_DESCR(ainput)->type_num, 0, 1);
-    if (avals == NULL) {
+    ni = PyArray_SIZE(mask);
+    if (ni != max_item) {
+        PyErr_SetString(PyExc_ValueError,
+                        "place: mask and data must be "
+                        "the same size");
         goto fail;
     }
-    numvals = PyArray_SIZE(avals);
-    nd = PyArray_NDIM(ainput);
-    input_data = PyArray_DATA(ainput);
-    mptr = PyArray_DATA(amask);
-    melsize = PyArray_DESCR(amask)->elsize;
-    vptr = PyArray_DATA(avals);
-    delsize = PyArray_DESCR(avals)->elsize;
-    zero = PyArray_Zero(amask);
-    if (zero == NULL) {
+
+    mask_data = PyArray_DATA(mask);
+    dtype = PyArray_DESCR(array);
+    Py_INCREF(dtype);
+    values = (PyArrayObject *)PyArray_FromAny(values0, dtype,
+                                    0, 0, NPY_ARRAY_CARRAY, NULL);
+    if (values == NULL) {
         goto fail;
     }
-    objarray = (PyArray_DESCR(ainput)->type_num == NPY_OBJECT);
 
-    if (!numvals) {
-        /* nothing to insert! fail unless none of mask is true */
-        const char *iter = mptr;
-        const char *const last = iter + PyArray_NBYTES(amask);
-        while (iter != last && !memcmp(iter, zero, melsize)) {
-            iter += melsize;
+    nv = PyArray_SIZE(values); /* zero if null array */
+    if (nv <= 0) {
```

Comment:
You could make this check part of the main loop, and spare yourself an extra iteration over mask.  It would be (the main loop) something like:

```C
npy_bool empty_values = 0;
for (i = 0; i < ni; i++) {
    if (mask_data[i]) {
        if (j >= nv) {
            if (nv == 0) {
                empty_values = 1;
                break;
            }
            j = 0
        }
        copyswap(dest + i*chunk, src + j*chunk, 0, array);
        j++;
    }
}
if (empty_values) {
    PyErr_SetString(PyExc_ValueError,
                    "Cannot insert from an empty array!");
}
```

Even if you decide to keep the emptiness check separate, testing for `j >= nv` only when a value is actually needed is probably a good idea.